### PR TITLE
Workflow to run performance tests using opensearch-benchmark

### PR DESCRIPTION
**Pull request by rishabh6788**
### Description
This PR adds support to run performance tests against OpenSearch cluster using opensearch-benchmark. The workflow is very similar to `perf-tests` with below mentioned updates and improvements:
* Uses public cdk infra package [opensearch-cluster-cdk](https://github.com/opensearch-project/opensearch-cluster-cdk) instead of private opensearch-infra repo. 
* Uses opensearch-benchmark for running and analyzing performance run tests instead of custom internal tool. 
* Supports single and as well as multi-node clusters. 
* Supports adding additional OS config from the command line. 

### Issues Resolved
https://github.com/opensearch-project/opensearch-benchmark/issues/102

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/test_workflow/benchmark_test/benchmark_test_cluster.py
Comment (dblock) [inline src/test_workflow/benchmark_test/benchmark_test_cluster.py:55]: Leftover print?

File: src/test_workflow/benchmark_test/benchmark_args.py
Comment (dblock) [inline src/test_workflow/benchmark_test/benchmark_args.py:55]: Are arguments elsewhere `camelCase`? I think not.

File: src/test_workflow/benchmark_test/benchmark_test_cluster.py
Comment (dblock) [inline src/test_workflow/benchmark_test/benchmark_test_cluster.py:57]: Why is this needed?

File: src/test_workflow/benchmark_test/benchmark_test_cluster.py
Comment (dblock) [inline src/test_workflow/benchmark_test/benchmark_test_cluster.py:69]: Use `f'...`

File: src/test_workflow/benchmark_test/benchmark_test_cluster.py
Comment (dblock) [inline src/test_workflow/benchmark_test/benchmark_test_cluster.py:79]: Use `logging` if you need to print something.

File: src/test_workflow/benchmark_test/benchmark_test_cluster.py
Comment (dblock) [inline src/test_workflow/benchmark_test/benchmark_test_cluster.py:134]: There's a lot of pass through params. I wonder whether we're better off not exposing all these command line options and only accepting a config file and passing it through here?

**Review by dblock — COMMENTED** [review]
Looks pretty clean! Some nits.

Please document this in .md as part of this PR?

File: src/test_workflow/benchmark_test/benchmark_test_cluster.py
Comment (rishabh6788) [reply src/test_workflow/benchmark_test/benchmark_test_cluster.py]: Removed.

File: src/test_workflow/benchmark_test/benchmark_args.py
Comment (rishabh6788) [reply src/test_workflow/benchmark_test/benchmark_args.py]: this was a miss, updated to use uniform formatting.

File: src/test_workflow/benchmark_test/benchmark_test_cluster.py
Comment (rishabh6788) [reply src/test_workflow/benchmark_test/benchmark_test_cluster.py:56]: It has more to do how opensearch-cluster-cdk accepts the json type parameter, the input has to be enclosed in single-quote in order to be properly parsed in the CDK Code [here](https://github.com/opensearch-project/opensearch-cluster-cdk/blob/main/lib/os-cluster-entrypoint.ts#L139-L147). 
So in the above code I am just making sure that if the parameter is of type `additional-config` make sure the value is enclosed in single-quotes. Hope that makes sense. 
@dblock

File: src/test_workflow/benchmark_test/benchmark_test_cluster.py
Comment (rishabh6788) [reply src/test_workflow/benchmark_test/benchmark_test_cluster.py]: Updated

File: src/test_workflow/benchmark_test/benchmark_test_cluster.py
Comment (rishabh6788) [reply src/test_workflow/benchmark_test/benchmark_test_cluster.py]: Not required, removed.

File: src/test_workflow/benchmark_test/benchmark_test_cluster.py
Comment (rishabh6788) [reply src/test_workflow/benchmark_test/benchmark_test_cluster.py:132]: Could you please give an example here, not sure if I understood the comment properly. 
@dblock

File: src/test_workflow/benchmark_test/benchmark_test_cluster.py
Comment (dblock) [reply src/test_workflow/benchmark_test/benchmark_test_cluster.py:56]: It seems weird. Is there a more generic version of this? For example if the value has quotes it should be single quoted?

File: src/run_benchmark_test.py
Comment (dblock) [inline src/run_benchmark_test.py:25]: Should a failure return a non-zero error code?

File: src/test_workflow/benchmark_test/benchmark_args.py
Comment (dblock) [inline src/test_workflow/benchmark_test/benchmark_args.py:71]: Another camelCase

File: src/test_workflow/benchmark_test/benchmark_test_cluster.py
Comment (dblock) [inline src/test_workflow/benchmark_test/benchmark_test_cluster.py:125]: Should be able to do str(self.args.insecure) and always get the right value of "true" or "false" - default value for insecure should be `false`

File: src/test_workflow/benchmark_test/benchmark_test_cluster.py
Comment (dblock) [reply src/test_workflow/benchmark_test/benchmark_test_cluster.py:132]: I think I'm saying that almost all `Count` arguments come from args. I don't know whether that's the right answer, but a config file that contained all these options could be passed as 1 single argument `--config foobar.yml`. Then this code doesn't need to know about `dataNodeCount` for example, that would be contained inside `foobar.yml`.

Comment (rishabh6788) [reply src/test_workflow/benchmark_test/benchmark_test_cluster.py:132]: The idea to keep them as args is that there may be cases where a user wants to try different cluster configurations for running benchmark in parallel with different node counts for each cluster. Another is the ML use case where a user wants to have ML node enabled. 
@dblock

File: src/run_benchmark_test.py
Comment (rishabh6788) [reply src/run_benchmark_test.py:25]: makes sense, will remove it.

File: src/test_workflow/benchmark_test/benchmark_args.py
Comment (rishabh6788) [reply src/test_workflow/benchmark_test/benchmark_args.py]: fixed it!

File: src/test_workflow/benchmark_test/benchmark_test_cluster.py
Comment (rishabh6788) [reply src/test_workflow/benchmark_test/benchmark_test_cluster.py:125]: I guess I resorted to this because the boolean in python is `True` or `False` but in typescript and many other languages it is `true` or `false`. I think better way to handle this would be:
`str(self.args.insecure).lower()`.

**Review by dblock — CHANGES_REQUESTED** [review]
You still have the `camelCase` variables all over the place, that's not the convention in the rest of the project. Example: https://github.com/opensearch-project/opensearch-build/blob/main/src/test_workflow/perf_test/perf_args.py#L21

File: src/test_workflow/benchmark_test/benchmark_test_cluster.py
Comment (dblock) [reply src/test_workflow/benchmark_test/benchmark_test_cluster.py:56]: Does quoting everything work, ie. doing `params_list.append(f" -c {key}=\'{value}\'")` all the time?

If not, add a TODO: fixme here or something like that with an explanation that this is a hack

Next, if you want to clean this up further,  you can write the loop with a `map` and a lambda that will make this a lot shorter. Finally, add role params that are `-c` into `params_dict` and let them be transformed with this code to avoid hard-coding strings. You'll still need the `--require-approval` and the `--plugin` params I guess. I'm ok to leave it as is, but consider doing it.

File: src/test_workflow/benchmark_test/benchmark_test_cluster.py
Comment (dblock) [inline src/test_workflow/benchmark_test/benchmark_test_cluster.py:71]: remove commented code

File: src/test_workflow/benchmark_test/benchmark_test_cluster.py
Comment (dblock) [reply src/test_workflow/benchmark_test/benchmark_test_cluster.py:132]: That's fine. I think we're just arguing command line params vs. a config file.

File: src/test_workflow/benchmark_test/benchmark_test_cluster.py
Comment (rishabh6788) [reply src/test_workflow/benchmark_test/benchmark_test_cluster.py:56]: Ack! Will add a TODO and take this up in the next iteration.

**Comment by rishabh6788** [comment]
> You still have the `camelCase` variables all over the place, that's not the convention in the rest of the project. Example: https://github.com/opensearch-project/opensearch-build/blob/main/src/test_workflow/perf_test/perf_args.py#L21

Got it! let me refactor everything and update accordingly. Too much coding in TS got me into this habit :-|

**Comment by rishabh6788** [comment]
> You still have the `camelCase` variables all over the place, that's not the convention in the rest of the project. Example: https://github.com/opensearch-project/opensearch-build/blob/main/src/test_workflow/perf_test/perf_args.py#L21

Fixed it. Also added the TODO to get rid of hack I put to quote the json input argument. 
@dblock

File: src/test_workflow/benchmark_test/benchmark_args.py
Comment (dblock) [inline src/test_workflow/benchmark_test/benchmark_args.py:57]: Put each argument such as `help` on its own line and it won't need to be wrapped like this.

File: src/test_workflow/test_jsonargs.py
Comment (dblock) [inline src/test_workflow/test_jsonargs.py:24]: Call this file `json_args.py` and this won't be needed. It's also the convention.

File: src/test_workflow/test_jsonargs.py
Comment (dblock) [inline src/test_workflow/test_jsonargs.py:18]: Stray `printf`, remove.

File: src/test_workflow/benchmark_test/benchmark_test_suite.py
Comment (dblock) [inline src/test_workflow/benchmark_test/benchmark_test_suite.py:44]: Looks messy, and uses an unnecessary local `base_command` variable.

```python
self.command = "docker run"
if args.benchmark_config:
   self.command += " -v ..."
self.command += "opensearchproject/opensearch-benchmark:latest execute_test:
...
```

**Review by dblock — CHANGES_REQUESTED** [review]
Needs documentation in .md files for these new tests.

Code: I'm down to nits. Address those and it's ready to merge.

**Comment by rishabh6788** [comment]
> Needs documentation in .md files for these new tests.
> 
> Code: I'm down to nits. Address those and it's ready to merge.

Added `README` for benchmarking test. Have a TODO for adding jenkins job details once it is ready. It is still WIP. 
Addressed the comments. 
@dblock

File: src/test_workflow/benchmark_test/benchmark_test_suite.py
Comment (rishabh6788) [reply src/test_workflow/benchmark_test/benchmark_test_suite.py]: Done!

File: src/test_workflow/benchmark_test/benchmark_test_suite.py
Comment (rishabh6788) [reply src/test_workflow/benchmark_test/benchmark_test_suite.py]: Thanks, done!

File: src/test_workflow/test_jsonargs.py
Comment (rishabh6788) [reply src/test_workflow/test_jsonargs.py]: Done!

File: src/test_workflow/test_jsonargs.py
Comment (rishabh6788) [reply src/test_workflow/test_jsonargs.py]: Makes sense, done!

File: src/test_workflow/benchmark_test/benchmark_args.py
Comment (rishabh6788) [reply src/test_workflow/benchmark_test/benchmark_args.py]: Done!

File: src/test_workflow/README.md
Comment (dblock) [inline src/test_workflow/README.md:177]: Update TOC

File: src/test_workflow/README.md
Comment (dblock) [inline src/test_workflow/README.md:179]: Don't line wrap, https://code.dblock.org/2021/06/07/to-wrap-or-not-to-wrap-in-markdown.html

File: src/test_workflow/README.md
Comment (dblock) [inline src/test_workflow/README.md:178]: test or tests?

you can remove "to run benchmark tests", that's what it does

File: src/test_workflow/README.md
Comment (dblock) [inline src/test_workflow/README.md:183]: I believe we don't abbreviate OS, spell OpenSearch.

File: src/test_workflow/README.md
Comment (dblock) [inline src/test_workflow/README.md:251]: Capitalize Like Other Titles

You can remove this for now since we don't have content.

**Review by dblock — CHANGES_REQUESTED** [review]
Minor stuff for the markdown please.

File: src/test_workflow/README.md
Comment (rishabh6788) [reply src/test_workflow/README.md:178]: Sorry about that, updated!

File: src/test_workflow/README.md
Comment (rishabh6788) [reply src/test_workflow/README.md]: Ack! fixed!

File: src/test_workflow/README.md
Comment (rishabh6788) [reply src/test_workflow/README.md]: Done!

File: src/test_workflow/README.md
Comment (rishabh6788) [reply src/test_workflow/README.md]: Updated!

File: src/test_workflow/README.md
Comment (rishabh6788) [reply src/test_workflow/README.md]: Removed for now. Will add once we have the job ready to be merged.

**Review by dblock — APPROVED** [review]

**Pull request merged**
